src/sentry/models/authenticator.py

```diff
@@ -0,0 +1,489 @@
+    """This interface sends OTP codes via text messages to the user."""
+    type = 2
+    interface_id = 'sms'
+    name = _('Text Message')
+    description = _('This authenticator sends you text messages for '
+                    'verification.  It\'s useful as a backup method '
+                    'or when you do not have a phone that supports '
+                    'an authenticator application.')
+    code_ttl = 45
+
+    @property
+    def is_available(self):
+        return sms_available()
+
+    def generate_new_config(self):
+        config = super(SmsInterface, self).generate_new_config()
+        config['phone_number'] = None
+        return config
+
+    def make_otp(self):
+        return TOTP(self.config['secret'], digits=6, interval=self.code_ttl,
+                    default_window=1)
+
+    def _get_phone_number(self):
+        return self.config['phone_number']
+
+    def _set_phone_number(self, value):
+        self.config['phone_number'] = value
+
+    phone_number = property(_get_phone_number, _set_phone_number)
+    del _get_phone_number, _set_phone_number
+
+    def activate(self, request):
+        if self.send_text(request=request):
+            return ActivationMessageResult(
+                _('A confirmation code was sent to your phone. '
+                  'It is valid for %d seconds.') % self.code_ttl)
+        return ActivationMessageResult(
+            _('Error: we failed to send a text message to you. You '
+              'can try again later or sign in with a different method.'),
+            type='error')
+
+    def send_text(self, for_enrollment=False, request=None):
+        ctx = {'code': self.make_otp().generate_otp()}
+
+        if for_enrollment:
+            text = _('You are about to set up two-factor authentication '
+                     'through text messages. Your confirmation code is '
+                     '%(code)s.')
+        else:
+            text = _('%(code)s is your Sentry authentication code.')
+
+        if request is not None:
+            text = u'%s\n\n%s' % (text, _('Requested from %(ip)s'))
+            ctx['ip'] = request.META['REMOTE_ADDR']
+
+        return send_sms(text % ctx, to=self.phone_number)
+
+
+@register_authenticator
+class U2fInterface(AuthenticatorInterface):
+    type = 3
+    interface_id = 'u2f'
+    name = _('U2F (Universal 2nd Factor)')
+    description = _('Authenticate with a U2F hardware device. This is a '
+                    'device like a Yubikey or something similar which '
+                    'supports FIDO\'s U2F specification. This also requires '
+                    'a browser which supports this system (like Google '
+                    'Chrome).')
+
+    @property
+    def u2f_app_id(self):
+        return options.get('system.url-prefix')
+
+    @property
+    def u2f_facets(self):
+        return [self.u2f_app_id]
+
+    @property
+    def is_available(self):
+        return self.u2f_app_id.startswith('https://')
```

Comment:
It's technically possible for `options.get()` to return `None`, in which case this is going to error.